### PR TITLE
Added implementation of inline tags and tag blocks

### PR DESCRIPTION
**Pull request by clecat**
This pull request contains the implementation of the tag extension system discussed [here](https://github.com/ocaml/omd/issues/190):

It allows the user to use self defined printers by using a new syntax:
For inline tags:
```
  This is an {!foo: inline tag}.
```
For tag blocks:
```
{!bar:
  This is a tag block
}
```

The html tag printers will be called with an object containing:
- the string of the tag (in the examples, it will be `bar`)
- in the case of an inline tag, the text (`inline tag` for the example)
- in the case of a tag block, the list of the blocks contained in the tag (here, a Paragraph).
- and finally the attributes

Nesting blocks inside a tag block is possible !

This implementation may need some corrections, I'm listening to all your comments and modifications requests :)

### CODE REVIEW

**Comment by nojb** [comment]
Cool, am a bit busy at the moment, but will take a look at this over the weekend. Thanks!

**Comment by clecat** [comment]
No problem, no need to rush :) Thanks for giving a look

**Comment by XVilka** [comment]
By the way, it is easier to send pull requests from a separate branch rather than from master - this way you can work on many different pull requests in parallel and rebase/merge/cherry-pick when required.

**Comment by clecat** [comment]
Thanks for the information @XVilka, I'm aware of that dont worry, and I do it from time to time, but laziness (and more often just omission) makes me push from master :p

File: src/sexp.ml [hunk 1]
```diff
@@ -58,6 +60,8 @@ let rec block = function
       List [Atom "html"; Atom s]
   | Link_def {label; destination; _} ->
       List [Atom "link-def"; Atom label; Atom destination]
+  | Tag_block {tag; content; _} ->
+      List [Atom "tag"; Atom tag; List (Atom "list-item" :: List.map block content)]
```
Comment (nojb) [inline src/sexp.ml]: Why is there a `list-item` here ? Shouldn't this be
```
List [Atom "tag"; Atom tag; List (List.map block content)]
```

File: src/parser.ml [hunk 2]
```diff
@@ -796,6 +824,10 @@ let parse s0 =
       unordered_list_item ind s
   | Some ('0'..'9') ->
       ordered_list_item ind s
+  | Some '{' ->
+      tag s
+  | Some '}' ->
```
Comment (nojb) [inline src/parser.ml]: Why is this necessary (returning `Ltag` when seeing `}`) ?

File: src/parser.ml [hunk 3]
```diff
@@ -767,6 +768,33 @@ let blank s =
...
+        | Some ':' ->
+            let _, a =
+              match Sub.head ~rev:() s with
+              | Some '}' ->
+                  attribute_string s
+              | _ ->
+                  s, Attributes.empty
+            in
+            let string_tag = Buffer.contents tag in
+            let string_tag = if string_tag = "" then "none" else string_tag in
```
Comment (nojb) [inline src/parser.ml]: I don't think we should return `"none"` if the tag string is empty. In fact, we should probably disallow empty tags. What do you think?

File: tests/tag/uppercase.ml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      | "capitalize" ->
+        let f i block =
+          p.block {p with text = print_upper_text; code = print_upper_code} b block;
+          if i < List.length t.content - 1 then
+            Buffer.add_char b '\n'
+        in
+        List.iteri f t.content
+      | _ -> Omd.default_printer.tag_block p b t
+    in
+    {Omd.default_printer with tag = print_tag; tag_block = print_tag_block}
```
Comment (nojb) [inline tests/tag/uppercase.ml]: Style: use punning whenever you can, by calling functions `tag`, `tag_block`, etc.

File: src/parser.ml (hunk 2 above)
Comment (nojb) [reply src/parser.ml]: Ah OK I understand, it is to signal the end of a tagged block. I would use `Lendtag` (to be defined).

File: src/sexp.ml (hunk 1 above)
Comment (clecat) [reply src/sexp.ml]: Okay my bad, I never use sexp before and simply refered to the use above, I'll take that into note

File: src/parser.ml (hunk 2 above)
Comment (clecat) [reply src/parser.ml]: Yes, better idea, I wanted to have a minimum of element in the variant, I'll change that

File: src/parser.ml (hunk 3 above)
Comment (clecat) [reply src/parser.ml]: Yes, better idea, thank you

File: tests/tag/uppercase.ml (hunk 4 above)
Comment (clecat) [reply tests/tag/uppercase.ml]: I corrected that and changed it in the html.ml file too: tell me if it's better now

**Comment by clecat** [comment]
About the tag syntax, I did not really think about any other, this one is good to me, but I dont have any preferences, so if you think another one could be better, I'll change it

**Comment by clecat** [comment]
Maybe we should change the syntax for something more common: We could copy que way code blocks are written with '~', something like:
```
~~~foo
bar
~~~
```
For blocks and
~~~
This is ~foo bar~
~~~
For inline tags (we may need a better way to specify the tag name for them)

What do you think of it ? people might be more comfortable with this as it looks like code blocks

**Comment by clecat** [comment]
My bad, '~' are already used for code blocks too, but if we find another good character, it could be a good idea

**Comment by pw374** [comment]
> Maybe we should change the syntax for something more common: We could copy que way code blocks are written with '~', something like:
> 
> ```
> ~~~foo
> bar
> ~~~
> ```
> 
> For blocks and
> 
> ```
> This is ~foo bar~
> ```
> 
> For inline tags (we may need a better way to specify the tag name for them)
> 
> What do you think of it ? people might be more comfortable with this as it looks like code blocks

> My bad, '~' are already used for code blocks too, but if we find another good character, it could be a good idea

Perhaps use `+` (it already has a meaning for itemized lists but it should be ambiguous in your case)

**Comment by clecat** [comment]
Let's go for that

**Comment by clecat** [comment]
Hello, I changed the syntax so it uses fenced code format instead.
I mostly copied the behaviour of fenced code so I would add a minimum of errors
Tell me what do you think of the current code (parsers may need some cleaning)

**Comment by clecat** [comment]
Any thoughts about it ? :D

**Comment by nojb** [comment]
> Any thoughts about it ? :D

Sorry for the delay. I will take a look in the next few days. Thanks!

**Comment by clecat** [comment]
No problem, I don't want to steal your time :p

**Comment by clecat** [comment]
Got any time to give it a look ?

**Comment by nojb** [comment]
> Got any time to give it a look ?

Hello @clecat, very sorry for the delay in looking at this, I have been swamped with other things. I will definitely look at this before monday. Thanks for the patience!

**Comment by nojb** [comment]
And yes, the parser part is tricky. I would like to clean that up and switch to something more understandable (maybe parser combinators?).

**Comment by clecat** [comment]
Thanks a lot for your review :D

I agree with you about the special case of the inline tags, it annoys me a lot, I should try to fing a best way to do that (maybe `+foo+this is the bar text+` in order to keep the tag clearly separated)

And I agree with the parser part too, I've finally understood how it works but a little bit of rework refactor or redesign could be nice, if I have the opportunity and you want me too, I could give it a look for another PR.

But for now do you agree with the PR ? Do you want me to change anything ? (I've also 2 smalls PRs backed for tables and definition list which could be merged before I start looking at the parser refactoring)

**Comment by nojb** [comment]
Seems OK except for the choice of syntax. I will merge this and let you take the lead in finding a better alternative later on. Refactoring the parser would be very welcome! Thanks!

**Comment by clecat** [comment]
No problem, I'll try to change it when I can

Just to be sure, what is bothering you is the inline special syntax ? Or the whole syntax

**Comment by nojb** [comment]
> Just to be sure, what is bothering you is the inline special syntax ? Or the whole syntax ?

In principle the inline syntax is a problematic. But more generally I think that we need to have some experience with this feature to know if this syntax is "right" or not.
